core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1409,6 +1426,77 @@ class KafkaApis(val requestChannel: RequestChannel,
       ))
   }
 
+  def handleDescribeTopicPartitionsRequest(request: RequestChannel.Request): Unit = {
+    metadataCache match {
+      case _: ZkMetadataCache =>
+        throw new InvalidRequestException("ZK cluster does not handle DescribeTopicPartitions request")
+      case _ =>
+    }
+    val KRaftMetadataCache = metadataCache.asInstanceOf[KRaftMetadataCache]
+
+    val describeTopicPartitionsRequest = request.body[DescribeTopicPartitionsRequest].data()
+    var topics = scala.collection.mutable.Set[String]()
+    describeTopicPartitionsRequest.topics().forEach(topic => topics.add(topic.name()))
+
+    val cursor = describeTopicPartitionsRequest.cursor()
+    val fetchAllTopics = topics.isEmpty
+    if (fetchAllTopics) {
+      metadataCache.getAllTopics().foreach(topic => topics.add(topic))
```

Comment:
If we're paginating through all topics and have a cursor, we can avoid gather only the desired topics during this O(n) loop through all topics. That would let us avoid another O(n) operation below to filter the undesired topics

Reply:
Now we loop twice. 
In the first time, we gather the topics lexicographically larger than the cursor.
In the second time, we iterate the topic list and query for their partitions.